File: two_pointer/triplets-with-smaller-sum.js
const triplet_with_smaller_sum = (arr, target_sum) => {
    let result = 0;
    arr.sort((a, b) => a - b);

    for (let i = 0; i < arr.length; i++) {
        let l = i + 1;
        let r = arr.length - 1;
        let target = target_sum - arr[i];

        while (l < r) {
            if (arr[l] + arr[r] >= target) {
                r--;
            } else {
                result += r - l;
                l++;
            }
        }
    }

    return result;
}

console.log(triplet_with_smaller_sum([-1, 0, 2, 3], 3)) // 2
console.log(triplet_with_smaller_sum([-1, 4, 2, 1, 3], 5)) // 4

File: two_pointer/python/triplets-with-smaller-sum.py
def triplet_with_smaller_sum(arr, target):
    count = 0

    arr.sort()

    for i in range(len(arr) - 2):
        current_target = target - arr[i]
        left, right = i + 1, len(arr) - 1
        while left < right:
            if arr[left] + arr[right] >= current_target:
                right -= 1
            else:
                count += right - left
                left += 1

    return count

def triplet_with_smaller_sum_ii(arr, target):
    triplets = []

    arr.sort()

    for i in range(len(arr) - 2):
        current_target = target - arr[i]
        left, right = i + 1, len(arr) - 1
        while left < right:
            if arr[left] + arr[right] >= current_target:
                right -= 1
            else:
                for t in range(right, left, -1):
                    triplets.append([arr[i], arr[left], arr[t]])
                left += 1

    return triplets

if __name__ == '__main__':
    print(triplet_with_smaller_sum([-1, 0, 2, 3], 3))
    print(triplet_with_smaller_sum([-1, 4, 2, 1, 3], 5))
    print(triplet_with_smaller_sum_ii([-1, 0, 2, 3], 3))
    print(triplet_with_smaller_sum_ii([-1, 4, 2, 1, 3], 5))
